File: everything/src/tools/get_resume_tool.js
import axios from 'axios';
import { wrapper as axiosCookiejarSupport } from 'axios-cookiejar-support';
import { CookieJar } from 'tough-cookie';
import dotenv from 'dotenv';
import path from 'path';
import { fileURLToPath } from 'url';

// Load environment variables from .env file in the root directory
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
dotenv.config({ path: path.resolve(__dirname, '../../../.env') }); // Go up three levels

// --- Configuration ---
const DEFAULT_RX_EMAIL = process.env.RX_RESUME_EMAIL || "emanuvaderland@gmail.com";
const DEFAULT_RX_PASSWORD = process.env.RX_RESUME_PASSWORD || "E4YSj9UiVuSB3uJ";
const DEFAULT_RX_BASE_URL = process.env.RX_RESUME_BASE_URL || "http://192.168.50.90:3050/api";

/**
 * Gets details of a specific resume by ID from Reactive Resume.
 *
 * @param {object} params - Parameters object.
 * @param {string} params.resume_id - Required. The UUID of the resume.
 * @param {string} [params.email] - Optional. Reactive Resume email (overrides env).
 * @param {string} [params.password] - Optional. Reactive Resume password (overrides env).
 * @param {string} [params.base_url] - Optional. Reactive Resume API base URL (overrides env).
 * @returns {Promise<object>} - Promise resolving to the resume details object or an error object.
 */
export async function getResume(params = {}) {
    // Create a cookie jar for this execution
    const cookieJar = new CookieJar();
    // Create an axios instance that uses the cookie jar
    const axiosInstance = axiosCookiejarSupport(axios.create({ jar: cookieJar }));

    let sessionCookie = null;

    try {
        // Validate required parameters
        if (!params.resume_id) {
            throw new Error("Resume ID (resume_id) is required");
        }

        // --- Apply Defaults and Environment Variables ---
        const resumeId = params.resume_id;
        const email = params.email || DEFAULT_RX_EMAIL;
        const password = params.password || DEFAULT_RX_PASSWORD;
        const baseUrl = params.base_url || DEFAULT_RX_BASE_URL;

        if (!email || !password || !baseUrl) {
            throw new Error("Missing required Reactive Resume authentication details (email, password, base_url) in params or environment variables.");
        }

        // --- Step 1: Authenticate with Reactive Resume ---
        const authUrl = `${baseUrl}/auth/login`;
        const authPayload = { identifier: email, password: password };
        const authHeaders = { 'Content-Type': 'application/json' };

        let sessionCookie = null; // Keep for potential debugging/other headers
        try {
            // Use the axios instance with cookie jar support
            const authResponse = await axiosInstance.post(authUrl, authPayload, { headers: authHeaders });
            if (authResponse.status !== 200) {
                throw new Error(`Authentication failed with status code ${authResponse.status}`);
            }
            // Cookie jar handles cookies automatically
            console.log("Authentication successful, cookie jar updated.");
            // Optionally extract for debugging
            const cookies = authResponse.headers['set-cookie'];
             if (cookies) {
                 sessionCookie = cookies.find(cookie => cookie.startsWith('connect.sid='))?.split(';')[0];
             }
        } catch (error) {
            const status = error.response?.status;
            const message = error.response?.data?.message || error.message;
            throw new Error(`Authentication failed${status ? ` with status ${status}` : ''}: ${message}`);
        }

        // --- Step 2: Get Resume Details ---
        const resumeUrl = `${baseUrl}/resume/${resumeId}`;
        // No need for manual cookie header, axiosInstance handles it
        const getHeaders = {};
        console.log(`Fetching resume details from: ${resumeUrl}`);

        try {
            // Use the axios instance with cookie jar support
            const resumeResponse = await axiosInstance.get(resumeUrl, { headers: getHeaders });

            if (resumeResponse.status !== 200) {
                throw new Error(`Failed to get resume with status code ${resumeResponse.status}. Response: ${JSON.stringify(resumeResponse.data)}`);
            }

            const resume = resumeResponse.data;

            // Format the response slightly for consistency if needed, or return directly
            // For now, returning the direct API response structure
            return resume;

        } catch (error) {
            const status = error.response?.status;
            const message = error.response?.data?.message || error.message;
            throw new Error(`Failed to fetch resume ${resumeId}${status ? ` with status ${status}` : ''}: ${message}`);
        }

    } catch (error) {
        console.error(`Error in getResume: ${error.message}`);
        return { error: `Error: ${error.message}` };
    }
}

// --- Tool Definition ---
export const getResumeTool = {
    name: "get_resume",
    description: "Gets the full details of a specific resume by ID from Reactive Resume. Uses env vars for auth defaults.",
    execute: getResume,
    inputSchema: {
        type: "object",
        properties: {
            resume_id: { type: "string", description: "ID of the resume to fetch (Required)" },
            email: { type: "string", description: "Reactive Resume email (Optional, overrides env)" },
            password: { type: "string", description: "Reactive Resume password (Optional, overrides env)" },
            base_url: { type: "string", description: "Reactive Resume API URL (Optional, overrides env)" }
        },
        required: ["resume_id"]
    }
};

// --- Example Usage (for testing) ---
// Note: Needs a valid resume_id from your Reactive Resume instance
async function runTest() {
     const isMain = import.meta.url.endsWith(path.basename(process.argv[1])) ||
                   import.meta.url.endsWith(path.basename(process.argv[1]) + '.js') ||
                   process.argv[1] === fileURLToPath(import.meta.url);

    if (isMain) {
        console.log("Running getResume test...");
        // --- !! Replace with a REAL resume ID from your instance !! ---
        const testResumeId = process.env.TEST_RESUME_ID || "replace-with-real-resume-id";
        // --- !! ----------------------------------------------- !! ---

        if (testResumeId === "replace-with-real-resume-id") {
             console.warn("Skipping test: TEST_RESUME_ID environment variable or default value is not set to a real ID.");
             return;
        }

        const testParams = {
            resume_id: testResumeId,
        };

        try {
            const result = await getResume(testParams);
            console.log("Get Resume Result:");
            console.log(JSON.stringify(result, null, 2));

            if (result.error) {
                 console.error("Test finished with error.");
            } else if (result.id === testResumeId) {
                 console.log("Test PASSED: Fetched resume ID matches requested ID.");
            } else {
                 console.warn("Test potentially failed: Fetched resume ID does not match requested ID (or result structure unexpected).");
            }

        } catch (error) {
            console.error("Get Resume Test Failed:");
            console.error(error);
        }
    }
}

runTest();

File: everything/src/source tools/get_resume_tool.py
import json
import requests
import os

def get_resume(params_json: str = "{}") -> str:
    """
    Get details of a specific resume by ID from Reactive Resume.
    
    Args:
        params_json: A JSON string containing parameters:
            {
                "resume_id": "uuid-of-resume",  # Required
                "email": "user@example.com",    # Optional, uses env var if not provided
                "password": "password",         # Optional, uses env var if not provided
                "base_url": "http://example.com/api"  # Optional, uses env var if not provided
            }
    
    Returns:
        str: JSON string containing the resume details or error message
    
    Example:
        >>> get_resume('{"resume_id": "123abc"}')
        "{"id": "123abc", "title": "My Resume", ...}"
    """
    try:
        # Parse parameters
        try:
            params = json.loads(params_json) if params_json else {}
        except json.JSONDecodeError:
            return "Error: Invalid JSON parameters"
        
        # Validate required parameters
        if not params.get("resume_id"):
            return "Error: Resume ID is required"
        
        # Configuration
        EMAIL = params.get("email") or os.environ.get("RX_RESUME_EMAIL", "emanuvaderland@gmail.com")
        PASSWORD = params.get("password") or os.environ.get("RX_RESUME_PASSWORD", "E4YSj9UiVuSB3uJ")
        BASE_URL = params.get("base_url") or os.environ.get("RX_RESUME_BASE_URL", "http://192.168.50.90:3050/api")
        
        # Step 1: Authenticate
        auth_url = f"{BASE_URL}/auth/login"
        auth_payload = {
            "identifier": EMAIL,
            "password": PASSWORD
        }
        auth_headers = {
            "Content-Type": "application/json"
        }
        
        session = requests.Session()
        auth_response = session.post(auth_url, headers=auth_headers, json=auth_payload)
        
        if auth_response.status_code != 200:
            return f"Error: Authentication failed with status code {auth_response.status_code}"
        
        # Step 2: Get resume details
        resume_url = f"{BASE_URL}/resume/{params['resume_id']}"
        resume_response = session.get(resume_url)
        
        if resume_response.status_code != 200:
            return f"Error: Failed to get resume with status code {resume_response.status_code}"
        
        resume = resume_response.json()
        
        # Format the response
        result = {
            "id": resume.get("id"),
            "title": resume.get("title"),
            "slug": resume.get("slug"),
            "visibility": resume.get("visibility"),
            "data": resume.get("data"),
            "created_at": resume.get("createdAt"),
            "updated_at": resume.get("updatedAt")
        }
        
        return json.dumps(result, indent=2)
    
    except requests.RequestException as e:
        return f"Error: Network error - {str(e)}"
    except Exception as e:
        return f"Error: Unexpected error - {str(e)}"

# For testing
if __name__ == "__main__":
    # Example usage with a resume ID
    # Get the first resume ID from list_resumes_tool
    try:
        import list_resumes_tool
        resumes_json = list_resumes_tool.list_resumes()
        resumes_data = json.loads(resumes_json)
        
        if resumes_data.get("count", 0) > 0 and "resumes" in resumes_data:
            first_resume_id = resumes_data["resumes"][0]["id"]
            test_params = json.dumps({"resume_id": first_resume_id})
            print(f"Testing with resume ID: {first_resume_id}")
        else:
            print("No resumes found. Using example resume ID.")
            test_params = json.dumps({"resume_id": "your-resume-id-here"})
    except Exception as e:
        print(f"Error getting resume list: {str(e)}")
        test_params = json.dumps({"resume_id": "your-resume-id-here"})
    
    result = get_resume(test_params)
    print(result)